src/rred_reports/dataclass_function.py

```diff
@@ -0,0 +1,189 @@
+    """Create a nested dataframe from an excel file
+
+    Args:
+        file (Path): Path object pointing to excel file
+
+    Returns:
+        pd.DataFrame: Nested pandas DataFrame
+    """
+    full_data = pd.read_excel(file)
+
+    def clmnlist(i):
+        return list(full_data.iloc[:, i])
+
+    all_schools_df = School(clmnlist(6), clmnlist(5), clmnlist(4), clmnlist(3))
+    all_schools_df = all_schools_df.drop_duplicates()  # pylint: disable=E1101
+
+    teach_df = Teacher(clmnlist(1), clmnlist(2), clmnlist(6))
+    teach_df = teach_df.drop_duplicates()  # pylint: disable=E1101
+
+    drop_cols = list(all_schools_df.columns.values)
+    drop_cols.append(teach_df.columns.values[1])
+    df_slimmed = full_data.drop(columns=drop_cols)
+    school_teacher = []
+    school_dict = all_schools_df.to_dict("records")
+    for school in school_dict:
+        school_df = pd.DataFrame.from_dict([school])
+        school_teacher.append(pd.merge(school_df["school_id"], teach_df, how="inner", on="school_id"))
+
+    teacher_pupils = []
+    for teacher in school_teacher:
+        pupil_df = pd.merge(teacher["rred_user_id"], df_slimmed, how="inner", on="rred_user_id")
+        teacher_pupils.append(pupil_df)
+
+    school_teacher_df = pd.DataFrame({"school_teacher": school_teacher})
+    teacher_pupil_df = pd.DataFrame({"teacher_pupil": teacher_pupils})
+
+    return pd.DataFrame({"dfs": [school_teacher_df, teacher_pupil_df]}, index=["school_teacher", "teacher_pupil"])
+
+
+def parse_nested_dataframe(nested_df: pd.DataFrame) -> dict[pd.DataFrame]:
+    """Helper function to parse the nested dataframe
+    Removes a single layer of nesting
+
+    Args:
+        nested_df (pd.DataFrame): Nested pandas DataFrame
+    Returns:
+        dict[pd.DataFrame]: _description_
+    """
+
+    school_teacher_df = nested_df.loc["school_teacher"].iloc[0]["school_teacher"]
+    teacher_pupil_df = nested_df.loc["teacher_pupil"].iloc[0]["teacher_pupil"]
+
+    return {"school_teacher": school_teacher_df, "teacher_pupil": teacher_pupil_df}
+
+
+def print_nested_dataframe_contents(nested_df: pd.DataFrame) -> None:
+    """Print the contents of a nested dataframe
+    Mainly for convenience/debugging. Should be replaced with
+    some proper logging if we're going to use this going forward.
+
+    Args:
+        nested_df (pd.DataFrame): Nested pandas DataFrame
+    """
+
+    for dataframe in nested_df:
+        logger.info(dataframe)
+
+
+def main():
+    """Entrypoint for generating nested dataframe"""
+    file_path = Path(__file__).resolve().parents[2] / "input" / "templates" / "example_dataset.xlsx"
+    full_nested_df = createdf(file_path)
+
+    unravelled_data = parse_nested_dataframe(full_nested_df)
+
+    school_teacher_df = unravelled_data["school_teacher"]
+    teacher_pupil_df = unravelled_data["teacher_pupil"]
+
+    print_nested_dataframe_contents(school_teacher_df)
+    print_nested_dataframe_contents(teacher_pupil_df)
+
+    return full_nested_df

```

Comment:
this would be great to have moved to a test, and add the example_dataset. Maybe adding some asserts at the end (doesn't need to be extensive) to make sure that the dimensions the dataframes are right and that a couple of the values are correct.